Add maker command to generate data providers/persisters

antograssiot opened the pull request:
<!-- Please update this template with something that matches your PR -->
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tickets       | fixes #3849
| License       | MIT
| Doc PR        | no yet

I had this ready since a year probably for discussion

soyuka commented:
This is a really good start thanks! Maybe some improvements for later:
 - generate the support methods by asking what entity this data provider / persister targets
 - DataProvider: choose if you want to do collections, items or both (not sure if this is possible)

antograssiot commented:
I'll look into this soon

dunglas commented on src/Bridge/Symfony/Maker/MakeDataPersister.php:
```diff
@@ -0,0 +1,65 @@
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+
+class MakeDataPersister extends AbstractMaker
```
```suggestion
final class MakeDataPersister extends AbstractMaker
```

dunglas commented on src/Bridge/Symfony/Maker/MakeDataPersister.php:
```diff
@@ -0,0 +1,65 @@
+    public function configureCommand(Command $command, InputConfiguration $inputConfig)
+    {
+        $command
+            ->setDescription('Creates a new class to persists data')
```
```suggestion
            ->setDescription('Creates an API Platform data persister')
```

dunglas commented on src/Bridge/Symfony/Maker/MakeDataPersister.php:
```diff
@@ -0,0 +1,65 @@
+        $this->writeSuccessMessage($io);
+        $io->text([
+            'Next: Open your new data persister class and start customizing it.',
+            'Find the documentation at <fg=yellow>https://api-platform.com/docs/core/data-persisters/#data-persisters</>',
```
```suggestion
            'Find the documentation at <fg=yellow>https://api-platform.com/docs/core/data-persisters/</>',
```

dunglas commented on src/Bridge/Symfony/Maker/MakeDataProvider.php:
```diff
@@ -0,0 +1,65 @@
+    public function configureCommand(Command $command, InputConfiguration $inputConfig)
+    {
+        $command
+            ->setDescription('Creates a new class to provide data')
```
```suggestion
            ->setDescription('Creates an API Platform data povider')
```

dunglas commented on src/Bridge/Symfony/Maker/MakeDataProvider.php:
```diff
@@ -0,0 +1,65 @@
+        $this->writeSuccessMessage($io);
+        $io->text([
+            'Next: Open your new data provider class and start customizing it.',
+            'Find the documentation at <fg=yellow>https://api-platform.com/docs/core/data-providers/#data-providers</>',
```
```suggestion
            'Find the documentation at <fg=yellow>https://api-platform.com/docs/core/data-providers/</>',
```

dunglas commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataProvider.tpl.php:
```diff
@@ -0,0 +1,35 @@
+    /**
+    * {@inheritdoc}
+    */
+    public function getCollection(string $resourceClass, string $operationName = null, array $context = [])
```
```suggestion
    public function getCollection(string $resourceClass, string $operationName = null, array $context = []): iterable
```

dunglas commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataProvider.tpl.php:
```diff
@@ -0,0 +1,35 @@
+    /**
+    * {@inheritdoc}
+    */
+    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = [])
```
```suggestion
    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = []): object
```

Or maybe even the name of the entity?

dunglas reviewed (changes requested):
This looks great!

weaverryan commented:
I love this idea!

What about asking a question related to decoration? I commonly want a data persister or data provider simply so that I can decorate the core persister/provider to add some functionality - this is mostly useful with entities. The question could be:

> Will this data persister be used for an entity class?

(if yes):

> Choose which entity:

A) Foo
B) Bar
C) Other entities...

And then... I think we wouldn't even need to ask them if they want to "decorate"/call the core persister... I would think that if you have an entity then you *always* would.

Or, another option would be to ask:

> Choose which API Resource class this data persister will be used for?

A) Foo
B) Bar
C) ... list all other API Resource classes
D) None/other (in case the user is doing something super custom)

Then, if the class is an entity, we decorate the core entity persister.

It definitely adds significant complexity... but it would also be very nice, so I wanted to at least mention the idea :).

Thanks!

antograssiot commented:
I've improved the initial proposal by including @soyuka and @dunglas proposals

You can now choose in data providers it you only want to generate item or collection data providers
I'va also added a new question regarding the target resource class with auto-completion

dunglas commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataPersister.tpl.php:
```diff
@@ -0,0 +1,47 @@
+    * {@inheritdoc}
+    */
+    public function persist($data, array $context = [])<?php
+   if (null !==$resource_class) {
```
```suggestion
   if (null !== $resource_class) {
```

dunglas commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataProvider.tpl.php:
```diff
@@ -0,0 +1,52 @@
+use <?= $resource_full_class_name ?>;
+<?php endif; ?>
+
+final class <?= $class_name ?> implements <?= ($generate_collection ? 'ContextAwareCollectionDataProviderInterface, ' : '') . ($generate_item ? 'ItemDataProviderInterface, ' : '')?>RestrictedDataProviderInterface
```
`implode()`?

dunglas commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataPersister.tpl.php:
```diff
@@ -0,0 +1,47 @@
+    /**
+    * {@inheritdoc}
+    */
+    public function remove($data, array $context = [])
```
maybe ```suggestion
    public function remove($data, array $context = []): void
```

antograssiot commented:
updated as requested, I've also improve code coverage and rebased @dunglas 
I like @weaverryan suggestion on the decoration but it might fit in a future PR, I'm not sure when I'll have time to work on this, probably not before Christmas holidays

alanpoulain commented on src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php:
```diff
@@ -606,6 +608,16 @@ private function addDefaultsSection(ArrayNodeDefinition $rootNode): void
+    {
+        $rootNode
+            ->children()
+            ->arrayNode('maker')
+            ->{class_exists(MakerBundle::class) ? 'canBeDisabled' : 'canBeEnabled'}()
+            ->end()
```
```suggestion
                ->arrayNode('maker')
                    ->{class_exists(MakerBundle::class) ? 'canBeDisabled' : 'canBeEnabled'}()
                ->end()
```

alanpoulain commented on src/Bridge/Symfony/Maker/MakeDataPersister.php:
```diff
@@ -0,0 +1,109 @@
+    {
+        return 'make:data-persister';
+    }
+
```
You should add the `getCommandDescription` static method instead of a call to `setDescription` (seems new to make it lazy).

alanpoulain commented on src/Bridge/Symfony/Maker/MakeDataProvider.php:
```diff
@@ -0,0 +1,124 @@
+            ->addOption('collection-only', null, InputOption::VALUE_NONE, 'Generate only a collection data provider')
+            ->setHelp(file_get_contents(__DIR__.'/Resources/help/MakeDataProvider.txt'));
+
+        $inputConfig->setArgumentAsNonInteractive('name');
```
Why not using the automatic prompt for the name?

alanpoulain commented on src/Bridge/Symfony/Maker/Resources/help/MakeDataPersister.txt:
```diff
@@ -0,0 +1,5 @@
+The <info>%command.name%</info> command generates a new data persister class.
```
```suggestion
The <info>%command.name%</info> command generates an API Platform data persister class.
```

alanpoulain commented on src/Bridge/Symfony/Maker/Resources/help/MakeDataProvider.txt:
```diff
@@ -0,0 +1,5 @@
+The <info>%command.name%</info> command generates a new data provider class.
```
```suggestion
The <info>%command.name%</info> command generates an API Platform data provider class.
```

alanpoulain commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataPersister.tpl.php:
```diff
@@ -0,0 +1,47 @@
+<?= "<?php\n" ?>
+
```
No strict type?

alanpoulain commented on src/Bridge/Symfony/Maker/Resources/skeleton/DataPersister.tpl.php:
```diff
@@ -0,0 +1,47 @@
+use <?= $resource_full_class_name ?>;
+<?php endif; ?>
+
+final class <?= $class_name ?> implements ContextAwareDataPersisterInterface
```
Shouldn't we add the `ResumableDataPersisterInterface` too by default?

alanpoulain commented on tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php:
```diff
@@ -1265,6 +1265,9 @@ private function getBaseContainerBuilderProphecy(array $doctrineIntegrationsToLo
             'api_platform.listener.view.validate_query_parameters',
+            'api_platform.maker.command.data_persister',
+            'api_platform.maker.command.data_provider',
+            'api_platform.mercure.listener.response.add_link_header',
```
```suggestion
```

alanpoulain commented:
I love it! :heart_eyes:
I think we can merge it as is and improve it later (for instance with @weaverryan's comment) later.
WDYT @antograssiot @api-platform/core-team?

soyuka commented:
@alanpoulain feel free to continue working on this branch as @antograssiot doesn't have much time to continue!

alanpoulain commented:
Thank you very much @antograssiot!
